src/formats/SMI.cpp

```diff
@@ -0,0 +1,705 @@
+            new_atom.set_charge(new_atom.charge() - count_or_class);
+            break;
+        case '+':
+            count_or_class = read_number(smiles, ++i);
+            count_or_class = count_or_class == 0.0 ? 1.0 : count_or_class;
+            new_atom.set_charge(new_atom.charge() + count_or_class);
+            break;
+        case ':':
+            count_or_class = read_number(smiles, ++i);
+            new_atom.set("smiles_class", count_or_class);
+            break;
+        case ']':
+            --i;
+            break;
+        case '@':
+            // Set the direction of the chirality
+            if (i + 1 < smiles.size() && smiles[i + 1] == '@') {
+                chirality_type = "CW";
+                ++i;
+            } else if (i + 2 < smiles.size() &&
+                       chirality_tags.count(smiles.substr(i + 1, 2)) != 0) {
+                // Set the type of the chirality
+                chirality_type += " " + smiles.substr(i + 1, 2);
+                i += 2;
+                chirality_type += std::to_string(read_number(smiles, ++i));
+            }
+            new_atom.set("chirality", chirality_type);
+            break;
+        default:
+            warning("SMI Reader", "unknown property code: '{}'", smiles[i]);
+            break;
+        }
+        ++i;
+    }
+}
+
+void SMIFormat::read(Frame& frame) {
+    // Initialize all the reading variables
+    branch_point_ = std::stack<size_t>(); 
+    rings_ids_.clear();
+    mol_vector_.clear();
+    current_atom_ = 0;
+    previous_atom_ = 0;
+    current_bond_order_ = Bond::SINGLE;
+    first_atom_ = true;
+
+    Topology topol;
+    size_t groupid = 1;
+    mol_vector_.push_back(Residue("group " + std::to_string(groupid)));
+
+    auto smiles = trim(file_->readline());
+
+    auto check_ring =
+    [&](size_t ring_id) {
+        auto ring_lookup = rings_ids_.find(ring_id);
+
+        if (ring_lookup == rings_ids_.end()) {
+            rings_ids_.insert({ring_id, {previous_atom_, current_bond_order_}});
+            current_bond_order_ = Bond::SINGLE;
+            return;
+        }
+
+        // Deviation from the standard, technically bond orders need to be equal,
+        // but we will accept the stored order if the current order is single.
+        // This is common practice
+        topol.add_bond(previous_atom_,
+            ring_lookup->second.first,
+            current_bond_order_ == Bond::SINGLE?
+                ring_lookup->second.second :
+                current_bond_order_
+        );
+        rings_ids_.erase(ring_lookup);
+
+        current_bond_order_ = Bond::SINGLE;
+    };
+
+    size_t i;
+    for (i = 0; i < smiles.size(); ++i) {
+
+        if (std::isblank(smiles[i]) != 0) {
+            break;
+        }
+
+        if (smiles[i] == '[') {
+            // TODO: Don't pass the full string, instead pass an iterator
+            // or maybe a substring till the closing ']'
```

Comment:
we really need a C++11 port of std::string_view =/

Reply:
Agreed